public/app/core/utils/CorsSharedWorker.ts

```diff
@@ -0,0 +1,52 @@
+// Almost identical to CorsWorker.ts. Main difference being it allows loading a SharedWorker if browser supports it
+
+export function sharedWorkersSupported() {
+  return typeof window.SharedWorker !== 'undefined';
+}
+
+class SharedWorkerNotSupported implements SharedWorker {
+  onerror() {}
+  // @ts-ignore
+  readonly port: MessagePort;
+  dispatchEvent(): boolean {
+    return false;
+  }
+  addEventListener(): void {}
+  removeEventListener(): void {}
+}
+
+/**
+ * The base class is created dynamically here to allow later on syntax like:
+ * import { CorsSharedWorker as SharedWorker } from '../utils/CorsSharedWorker';
+ * And also take into account cases where SharedWorked is not available in the browser (mainly mobile browsers)
+ *
+ * It's important to use: new SharedWorker(...) syntax instead of new CorsSharedWorker(...) due to how web-workers
+ * are processing workers syntax (more details https://webpack.js.org/guides/web-workers/)
+ */
+const BaseSharedWorkerClass = sharedWorkersSupported() ? window.SharedWorker : SharedWorkerNotSupported;
+
+export class CorsSharedWorker extends BaseSharedWorkerClass {
+  constructor(url: URL, options?: WorkerOptions) {
+    if (!sharedWorkersSupported()) {
+      return;
+    }
+    // by default, worker inherits HTML document's location and pathname which leads to wrong public path value
+    // the CorsWorkerPlugin will override it with the value based on the initial worker chunk, ie.
+    //    initial worker chunk: http://host.com/cdn/scripts/worker-123.js
+    //    resulting public path: http://host.com/cdn/scripts
+
+    const scriptUrl = url.toString();
+    const urlParts = scriptUrl.split('/');
+    urlParts.pop();
+    const scriptsBasePathUrl = `${urlParts.join('/')}/`;
+
+    const importScripts = `importScripts('${scriptUrl}');`;
+    const objectURL = URL.createObjectURL(
```

Comment:
Hmm this is a little concerning. Generating a blob and `objectURL` of type `application/javascript` from what looks like arbitrary urls seems like a risk. Do we have full control of the input urls and are sure they are sanitised? Where is the `objectURL` going to be executed?

Reply:
This is exactly what we already do in `CorsWorker` class and is a workaround to load web workers from external domains. When calling `new Worker(...)` the browser [uses same-origin mode](https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-classic-worker-script). We create proxy/shell script (a blob) that loads the actual worker code via `importScripts()` which [uses CORS/CSP like a regular fetch](https://html.spec.whatwg.org/multipage/workers.html#worker-processing-model:fetching-scripts-perform-fetch)